lib/Targets/AIETargetAirbin.cpp

```diff
@@ -0,0 +1,1250 @@
+//===- AIETargetAirbin.cpp --------------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2021 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/None.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <ext/stdio_filebuf.h>
+#include <fcntl.h> // open
+#include <gelf.h>
+#include <iostream>
+#include <sstream>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h> // read
+#include <utility>  // pair
+#include <vector>
+
+#include "aie/Dialect/AIE/AIENetlistAnalysis.h"
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "aie/Dialect/AIEX/IR/AIEXDialect.h"
+#include "libelf.h"
+
+// Marks a particular code path as unfinished.
+#define TODO assert(false)
+
+// Marks a particular code path as unused in normal execution.
+#define UNREACHABLE assert(false)
+
+namespace xilinx {
+namespace AIE {
+
+enum {
+  SEC_IDX_NULL,
+  SEC_IDX_SSMAST,
+  SEC_IDX_SSSLVE,
+  SEC_IDX_SSPCKT,
+  SEC_IDX_SDMA_BD,
+  SEC_IDX_SHMMUX,
+  SEC_IDX_SDMA_CTL,
+  SEC_IDX_PRGM_MEM,
+  SEC_IDX_TDMA_BD,
+  SEC_IDX_TDMA_CTL,
+  SEC_IDX_DEPRECATED,
+  SEC_IDX_DATA_MEM,
+  SEC_IDX_MAX
+};
+
+static constexpr auto disable = 0u;
+static constexpr auto enable = 1u;
+static constexpr auto DMA_S2MM_CHANNEL_COUNT = 2u;
+static constexpr auto DMA_MM2S_CHANNEL_COUNT = 2u;
+
+static constexpr auto TILE_ADDR_OFF_WIDTH = 18u;
+
+static constexpr auto TILE_ADDR_ROW_SHIFT = TILE_ADDR_OFF_WIDTH;
+static constexpr auto TILE_ADDR_ROW_WIDTH = 5u;
+
+static constexpr auto TILE_ADDR_COL_SHIFT =
+    TILE_ADDR_ROW_SHIFT + TILE_ADDR_ROW_WIDTH;
+static constexpr auto TILE_ADDR_COL_WIDTH = 7u;
+
+static constexpr auto TILE_ADDR_ARR_SHIFT =
+    TILE_ADDR_COL_SHIFT + TILE_ADDR_COL_WIDTH;
+
+#define DMA_BD_COUNT 16
+
+#define REG_DMA_BD_BLOCK_SIZE 0x20
+#define REG_DMA_ADDR_A_BD(_idx) (0x1D000 + ((_idx)*REG_DMA_BD_BLOCK_SIZE) + 0x0)
```

Comment:
What about just doing C++14?
```suggestion
auto reg_dma_addr_a_bd = [](auto _idx) { return 0x1D000 + (_idx*REG_DMA_BD_BLOCK_SIZE) + 0x0; };
```
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es31-dont-use-macros-for-constants-or-functions
With C++20 you can use functions with `auto` instead of a lambda, but I guess you are stuck to C++17 in this project.

Reply:
The last time I used C++ seriously was around 2005, so I really don't know any of these new features :-)
I'm happy to try all your suggestions.

Reply:
For example I remember https://www.youtube.com/watch?v=ZLlGO102tVw which was interesting and I mentioned on Xilinx Slack `#cpp` along others.

Reply:
done